### PR TITLE
[RPA-1259] Allow for adding new collections

### PR DESCRIPTION
**Pull request by star-in-the-jar**
## Description
image
image

migration worked successfully
image

image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics-orchestrator/src/main/java/com/runbotics/domain/ProcessCollection.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
+package com.runbotics.domain;
+
+import jdk.jfr.BooleanFlag;
```
Comment (preludium) [inline runbotics-orchestrator/src/main/java/com/runbotics/domain/ProcessCollection.java]: whats this import?

File: runbotics-orchestrator/src/main/java/com/runbotics/repository/ProcessCollectionRepository.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.UUID;
+
+@SuppressWarnings("unused")
```
Comment (preludium) [inline runbotics-orchestrator/src/main/java/com/runbotics/repository/ProcessCollectionRepository.java]: seems like you use this repository so remove this warning

File: runbotics-orchestrator/src/main/java/com/runbotics/security/FeatureKeyConstants.java [hunk 3]
```diff
@@ -65,5 +65,9 @@ public final class FeatureKeyConstants {
 
     public static final String BASIC_USER_READ = "BASIC_USER_READ";
 
+    public static final String PROCESS_COLLECTION_READ = "PROCESS_COLLECTION_READ";
+
+    public static final String PROCESS_COLLECTION_ADD = "PROCESS_COLLECTION_ADD";
```
Comment (preludium) [inline runbotics-orchestrator/src/main/java/com/runbotics/security/FeatureKeyConstants.java]: add all possible already e.g. delete and edit

File: runbotics/runbotics-common/src/collections/process-collections.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { ProcessCollection } from "../model";
+
+export const COLLECTION_ID_PARAM = 'collectionId';
+
+export type ProcessCollectionKeys = { [Key in keyof ProcessCollection]: Key }[keyof ProcessCollection];
```
Comment (preludium) [inline runbotics/runbotics-common/src/collections/process-collections.ts:5]: what is it for?

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.slice.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    childrenProcesses: {
+        isLoading: false,
+        list: null,
+        byPage: null,
+    },
+    childrenCollections: {
+        isLoading: false,
+        list: null,
+    }
+};
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.slice.ts]: let's discuss it

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.extraReducers.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+const buildProcessCollectionExtraReducers = (builder: ActionReducerMapBuilder<ProcessCollectionState>) => {
+    builder
+        // CREATE ONE
+        .addCase(createOne.pending, (state) => {
+            state.childrenCollections.isLoading = true;
+        })
+        .addCase(createOne.fulfilled, (state) => {
+            state.childrenCollections.isLoading = false;
+        })
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.extraReducers.ts]: ideally, here you should update global state and you don't need to fetch all collections once again

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.extraReducers.ts (hunk 6 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.extraReducers.ts]: thanks for pointing that out

Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.extraReducers.ts]: done & testes

File: runbotics-orchestrator/src/main/java/com/runbotics/domain/ProcessCollection.java (hunk 1 above)
Comment (star-in-the-jar) [reply runbotics-orchestrator/src/main/java/com/runbotics/domain/ProcessCollection.java]: so removed it

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.slice.ts (hunk 5 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.slice.ts]: I'm going to be able to discuss this on friday 1p.m.

File: runbotics/runbotics-common/src/collections/process-collections.ts (hunk 4 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-common/src/collections/process-collections.ts:5]: There's universal function for handling every change in ModifyDialog form. This type is used to check if the property belongs to ProcessCollection object
image

File: runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessCollectionResource.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        ProcessCollectionQueryService processCollectionQueryService,
+        ProcessCollectionRepository processCollectionRepository
+    ) {
+        this.processCollectionService = processCollectionService;
+        this.processCollectionRepository = processCollectionRepository;
+        this.processCollectionQueryService = processCollectionQueryService;
+    }
+
+    @PreAuthorize("@securityService.checkFeatureKeyAccess('" + FeatureKeyConstants.BOT_COLLECTION_ADD + "')")
+    @PostMapping("process-collection")
```
Comment (preludium) [inline runbotics-orchestrator/src/main/java/com/runbotics/web/rest/ProcessCollectionResource.java]: wrong feature key

File: runbotics-orchestrator/src/main/resources/config/liquibase/changelog/20240123003633_changelog.xml
Comment (preludium) [inline runbotics-orchestrator/src/main/resources/config/liquibase/changelog/20240123003633_changelog.xml:1]: I believe that its already on develop brach. Maybe some issues appeard due to history changes on develop branch

Comment (preludium) [reply runbotics-orchestrator/src/main/resources/config/liquibase/changelog/20240123003633_changelog.xml:1]: yup, you have stale commits. Create a new branch after CR fixes and cherry pick your commits

**Review by preludium — CHANGES_REQUESTED** [review]

**Review by star-in-the-jar — COMMENTED** [review]
Self reviewed

**Review by star-in-the-jar — COMMENTED** [review]
self reviewed

File: runbotics-orchestrator/src/main/resources/config/liquibase/changelog/20240202155800_changelog.xml
Comment (preludium) [inline runbotics-orchestrator/src/main/resources/config/liquibase/changelog/20240202155800_changelog.xml:1]: to be adjusted

File: runbotics/runbotics-orchestrator-ui/src/src-app/utils/text.ts [hunk 8]
```diff
@@ -1,14 +1,14 @@
...
+    lowerCaseRest?: boolean,
+    delimiter?: string | RegExp,
     join?: string
 }
 
 export const capitalizeFirstLetter = ({
     text,
     lowerCaseRest = false,
-    delimiter,
+    delimiter = ' ',
```
Comment (preludium) [inline runbotics/runbotics-orchestrator-ui/src/src-app/utils/text.ts]: use `\s` for better readability

**Review by preludium — APPROVED** [review]

File: runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.slice.ts (hunk 5 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessCollection/ProcessCollection.slice.ts]: discussed, done

File: runbotics-orchestrator/src/main/resources/config/liquibase/changelog/20240202155800_changelog.xml
Comment (star-in-the-jar) [reply runbotics-orchestrator/src/main/resources/config/liquibase/changelog/20240202155800_changelog.xml:1]: tested if works, done

File: runbotics/runbotics-orchestrator-ui/src/src-app/utils/text.ts (hunk 8 above)
Comment (star-in-the-jar) [reply runbotics/runbotics-orchestrator-ui/src/src-app/utils/text.ts]: done

**Pull request merged**